### PR TITLE
[CIS-2] UI tokens table view

### PR DESCRIPTION
**Pull request by milsawicki**
## Purpose
Adds a protype cell for tokens and tab bar view above the table view. 
## Changes
Simulator Screen Recording - iPhone 14 Pro - 2023-08-09 at 00 28 28

### CODE REVIEW

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsViewController.swift
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsViewController.swift:1]: Consider moving `setupTransactionsUI` up next to `setupAccountTokensUI`:
https://github.com/Concordium/concordium-reference-wallet-ios/pull/357#discussion_r1286868907

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsViewController.swift [hunk 1]
```diff
@@ -88,15 +87,17 @@ class AccountDetailsViewController: BaseViewController, AccountDetailsViewProtoc
...
     
     func setupAccountTokensUI() {
         accountTokensViewController = AccountTokensViewController.instantiate(fromStoryboard: "Account") { coder in
-            return AccountTokensViewController(coder: coder)
+            let vc = AccountTokensViewController(coder: coder)
+            vc?.showTokenDetails = self.presenter.userSelected(_:)
+            vc?.showManageView = self.presenter.showManageView
+            return vc
         }
         add(child: accountTokensViewController, inside: containerView)
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsViewController.swift:102]: I think there is some value in using the same patterns as the project already uses, even if we think they're needlessly verbose. In this case that would mean moving this to something like `AccountTokensFactory`.

At least worth considering...

**Review by bisgardo — COMMENTED** [review]
Only had time for a breif look.

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsPresenter.swift [hunk 2]
```diff
@@ -105,6 +108,13 @@ class AccountDetailsPresenter {
 }
 
 extension AccountDetailsPresenter: AccountDetailsPresenterProtocol {
+    func userSelected(_ token: Token) {
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountDetailsPresenter.swift:111]: `tokenSelected` or `userSelected(token:)`?

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift [hunk 3]
```diff
@@ -7,13 +7,100 @@
...
+import Combine
+struct Token {
+    var id: String = ""
+    var token: String = ""
+    var isSelected = false
+    var contractIndex: String = ""
+    var subIndex: String = ""
+    var isCCDToken = false
+    var symbol = ""
+    var atDisposal: Float = 0.0
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:19]: Why the default values?

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift [hunk 4]
```diff
@@ -7,13 +7,100 @@
 //
 
 import UIKit
+import Combine
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:10]:
```suggestion
import Combine

```

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift [hunk 5]
```diff
@@ -7,13 +7,100 @@
...
+    var isSelected = false
+    var contractIndex: String = ""
+    var subIndex: String = ""
+    var isCCDToken = false
+    var symbol = ""
+    var atDisposal: Float = 0.0
 
-class AccountTokensViewController: UIViewController, Storyboarded {
-    @IBOutlet weak var ourLabel: UILabel!
+    init(id: String, token: String, isSelected: Bool = false, contractIndex: String, subIndex: String, isCCDToken: Bool = false, symbol: String = "") {
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:21]: same here

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift [hunk 6]
```diff
@@ -7,13 +7,100 @@
...
+        .init(id: "", token: "wCCD", contractIndex: "0", subIndex: "0", symbol: "wCCD")
+    ]
+}
+
+class AccountTokensViewController: BaseViewController, Storyboarded {
+
+    enum Tabs: Int, CaseIterable {
+        case fungible
+        case collectibles
+        case manage
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:43]:
```suggestion
        case manage
        
```

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift (hunk 3 above)
Comment (milsawicki) [reply ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:20]: that were some leftovers 👍🏻

File: ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift [hunk 7]
```diff
@@ -7,13 +7,113 @@
...
+        .init(id: "", token: "wCCD", isSelected: false, contractIndex: "0", subIndex: "0", symbol: "wCCD")
+    ]
+}
+
+class AccountTokensViewFactory {
+    class func create(with presenter: AccountDetailsPresenterProtocol) -> AccountTokensViewController {
+        AccountTokensViewController.instantiate(fromStoryboard: "Account") { coder in
+            let vc = AccountTokensViewController(coder: coder)
+            vc?.showTokenDetails = presenter.userSelected(token:)
+            vc?.showManageView = presenter.showManageView
```
Comment (bisgardo) [inline ConcordiumWallet/Views/AccountsView/AccountDetails/AccountTokens/AccountTokensViewController.swift:44]: I would rather pass the presenter (possibly as a protocol) compared to capturing method refs like this. That way it'd be much easier to track where the methods come from.

**Review by bisgardo — APPROVED** [review]

**Pull request merged**